### DISCUSSION TITLE
Cleaner way to get nested data in API?

### DISCUSSION THREAD

**Discussion by michaelmichael** [solidusio/solidus#4906] [Troubleshooting] [answered]
I have a model, `Document`, that inherits from `Spree::Asset`. The model utilizes ActiveStorage for attaching related documents to a product: brochures or user manuals, etc. Products can have many Documents, and they can be attached to variants or the master product.

I'm trying to use the Solidus API to access these documents. I can do that in an initializer:

``` ruby
Spree::Api::Config.configure do |config|
  config.product_attributes.push(:variant_docs)
end
```

But it doesn't give me everything I want. I really want is the attachment's path, but that isn't included by default. I have to write a `path_to_file` method, overwrite `views/spree/api/products/_product.json.jbuilder`, and include something like this in the jbuilder template:

``` ruby
  json.variant_docs(product.variant_docs) do |doc|
    json.name(doc.name)
    json.path(doc.path_to_file)
  end
```

But I'm wary of overwriting the whole view just to get this data. Is there a cleaner way to do it?

**Comment by kennyadsl** [comment]
@michaelmichael I didn't try but, according to [jbuilder documentation](https://github.com/rails/jbuilder), you could create a `to_builder` method on your `Document`:

```ruby
class Document
  # ...
  def to_builder
    Jbuilder.new do |doc|
      json.name(doc.name)
      json.path(doc.path_to_file)
    end
  end
end
```

so that jbuilder knows how to render that attribute. Let me know if that works!

**Reply by michaelmichael** [reply]
Thanks @kennyadsl! Seems like it's a step in the right direction, but doesn't work for me. It still uses the `to_builder` method from `Spree::Asset` for some reason. I might try writing an alias method and including that in my product attributes.

**Comment by michaelmichael** [comment, marked as answer]
I've more or less solved this by prepending a `doc_builder` method on `Spree::Product` that maps over documents and returns an object containing the data I want.

``` ruby
module Spree::Product::AddDocBuilder
  def doc_builder
    return nil unless variant_docs.any?

   docs = []

    variant_docs.map do |doc|
        docs.push({
          name: doc.name,
          attachment: doc.path_to_file,
          position: doc.position,
          id: doc.id
        })
        return docs
    end
  end
end
```

Then I just include that in the API config initializer: 

``` ruby
Spree::Api::Config.configure do |config|
  config.product_attributes.push(…, :doc_builder)
end
```

Seems like jbuilder just formats it correctly!
